File: src/main/python/AoC2023_07.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 7
#

from __future__ import annotations

import sys
from collections import Counter
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST = """\
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
"""

JOKER = "J"
BEST = "AAAAA"


class Hand(NamedTuple):
    cards: str
    bid: int
    value: int
    strength: str
    value_with_jokers: int
    strength_with_jokers: str

    @classmethod
    def from_input(cls, line: str) -> Hand:
        def get_value(cards: str) -> int:
            mc = Counter(cards).most_common()
            return 2 * mc[0][1] + (mc[1][1] if len(mc) > 1 else 0)

        def with_jokers(cards: str) -> str:
            c = Counter(card for card in cards if card != JOKER)
            if not c:
                return BEST
            best = c.most_common()[0][0]
            return "".join((card if card != JOKER else best) for card in cards)

        cards, bid = line.split()
        value = get_value(cards)
        strength = cards.translate(str.maketrans("TJQKA", "BCDEF"))
        value_with_jokers = get_value(with_jokers(cards))
        strength_with_jokers = cards.translate(str.maketrans("TJQKA", "B0DEF"))
        return Hand(
            cards,
            int(bid),
            value,
            strength,
            value_with_jokers,
            strength_with_jokers,
        )


Input = list[Hand]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [Hand.from_input(line) for line in input_data]

    def _solve(
        self,
        hands: Input,
        order_func: Callable[[Hand], tuple[int, str]],
    ) -> int:
        return sum(
            i * hand.bid
            for i, hand in enumerate(sorted(hands, key=order_func), start=1)
        )

    def part_1(self, hands: Input) -> Output1:
        return self._solve(hands, lambda hand: (hand.value, hand.strength))

    def part_2(self, hands: Input) -> Output2:
        return self._solve(
            hands,
            lambda hand: (hand.value_with_jokers, hand.strength_with_jokers),
        )

    @aoc_samples(
        (
            ("part_1", TEST, 6440),
            ("part_2", TEST, 5905),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 7)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_07.java
import static com.github.pareronia.aoc.IterTools.enumerateFrom;
import static com.github.pareronia.aoc.Utils.toAString;
import static java.util.Comparator.comparing;

import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

import com.github.pareronia.aoc.Counter;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_07
        extends SolutionBase<List<AoC2023_07.Hand>, Integer, Integer> {
    
    private AoC2023_07(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_07 create() {
        return new AoC2023_07(false);
    }
    
    public static AoC2023_07 createDebug() {
        return new AoC2023_07(true);
    }
    
    @Override
    protected List<Hand> parseInput(final List<String> inputs) {
        return inputs.stream().map(Hand::fromInput).toList();
    }
    
    private int solve(final List<Hand> hands, final Comparator<Hand> order) {
        return enumerateFrom(1, hands.stream().sorted(order))
            .mapToInt(e -> e.index() * e.value().bid())
            .sum();
    }

    @Override
    public Integer solvePart1(final List<Hand> hands) {
        return solve(hands, Hand.comparator());
    }
    
    @Override
    public Integer solvePart2(final List<Hand> hands) {
        return solve(hands, Hand.comparatorWithJokers());
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "6440"),
        @Sample(method = "part2", input = TEST, expected = "5905"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_07.create().run();
    }
    
    record Hand(
        int bid,
        int value,
        String strength,
        int valueWithJokers,
        String strengthWithJokers
    ) {
        private static final Character JOKER = 'J';
        private static final String BEST = "AAAAA";
        
        public static Hand fromInput(final String line) {
            final Function<String, Integer> getValue = cards -> {
                final List<Counter.Entry<Character>> mc
                    = new Counter<>(Utils.asCharacterStream(cards)).mostCommon();
                return (int) (2 * mc.get(0).count() + (mc.size() > 1 ? mc.get(1).count() : 0));
            };
            final Function<String, String> withJokers = cards -> {
                final Counter<Character> c = new Counter<>(
                    Utils.asCharacterStream(cards).filter(ch -> ch != JOKER));
                if (c.isEmpty()) {
                    return BEST;
                }
                final Character best = c.mostCommon().get(0).value();
                return Utils.asCharacterStream(cards)
                    .map(ch -> ch == JOKER ? best : ch)
                    .collect(toAString());
            };
            
            final StringSplit split = StringOps.splitOnce(line, " ");
            final int bid = Integer.parseInt(split.right());
            final int value = getValue.apply(split.left());
            final String strength = StringOps.translate(split.left(), "TJQKA", "BCDEF");
            final int valueWithJokers = getValue.apply(withJokers.apply(split.left()));
            final String strengthWithJokers = StringOps.translate(split.left(), "TJQKA", "B0DEF");
            return new Hand(bid, value, strength, valueWithJokers, strengthWithJokers);
        }
        
        public static Comparator<Hand> comparator() {
            return comparing(Hand::value).thenComparing(comparing(Hand::strength));
        }

        public static Comparator<Hand> comparatorWithJokers() {
            return comparing(Hand::valueWithJokers)
                        .thenComparing(comparing(Hand::strengthWithJokers));
        }
    }

    private static final String TEST = """
            32T3K 765
            T55J5 684
            KK677 28
            KTJJT 220
            QQQJA 483
            """;
}
